samza-sql/src/main/java/org/apache/samza/sql/interfaces/SamzaRelConverter.java

```diff
@@ -0,0 +1,78 @@
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.interfaces;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.schema.TableFactory;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+
+
+/**
+ * Samza SQL application uses {@link SamzaRelConverter} to convert the Samza messages to relational message before
+ * it can be processed by the calcite engine.
+ * The {@link SamzaRelConverter} is configurable at a system level, So it is possible to configure different
+ * {@link SamzaRelConverter} for different systems.
+ */
+public interface SamzaRelConverter {
+
+  /**
+   * Converts the schema to relational schema. It is possible that there can be multiple tables associated
+   * within a single schema.
+   * @param systemName the system name
+   * @param streamName Name of the stream
+   * @return Map of table name and it's associated relational schemas.
+   */
+  TableFactory<Table> getRelationalSchema(String systemName, String streamName);
+
+  /**
+   * Converts the object to relational message corresponding to the tableName with relational schema.
+   * @param key Key corresponding to the message.
+   * @param message message object that needs to be converted.
+   * @param dataType relational schema associated to the table.
+   * @return Relational message extracted from the object.
+   */
+  SamzaSqlRelMessage convertToRelMessage(Object key, Object message, RelDataType dataType);
+
+  /**
+   * Extract key object from the relational message.
+   * @param systemName the system name
+   * @param streamName the stream name
+   * @param relMessage the rel message
+   * @return the object
+   */
+/*
+   * Extracts the key from the relational message.
+   * @param streamName the stream name
+   * @param schema the schema
+   * @param relMessage relational message from which the key needs
+   * @return key to be used while writing to the output stream.
+   */
+  Object extractKey(String systemName, String streamName, SamzaSqlRelMessage relMessage);
+
+  /**
+   * Convert the relational message to the output message.
+   * @param systemName the system name
+   * @param streamName the stream name
+   * @param relMessage relational message that needs to be converted.
+   * @return the object
+   */
+  Object convertToSamzaMessage(String systemName, String streamName, SamzaSqlRelMessage relMessage);
```

Comment:
Same here. Can we return a typed KV<K,V> here?

Not sure why we need systemName and streamName here. Seems it doesn't fit this interface.

Reply:
For most scenarios, conversion from the relational to a specific wire format (e.g. avro) will need access to the schema corresponding to the stream. Hence the need for system and Stream.